client/base_client.go

```diff
@@ -364,32 +477,38 @@ func (c *baseClient) updateURLs(members []*pdpb.Member) {
 	c.urls.Store(urls)
 	// Update the connection contexts when member changes if TSO Follower Proxy is enabled.
 	if c.option.getEnableTSOFollowerProxy() {
-		c.scheduleUpdateConnectionCtxs()
+		// Run callbacks to refelect the membership changes in the leader and followers.
+		for _, cb := range c.membersChangedCallbacks {
+			cb()
+		}
 	}
 	log.Info("[pd] update member urls", zap.Strings("old-urls", oldURLs), zap.Strings("new-urls", urls))
 }
 
-func (c *baseClient) switchLeader(addrs []string) error {
+func (c *pdBaseClient) switchLeader(addrs []string) error {
 	// FIXME: How to safely compare leader urls? For now, only allows one client url.
 	addr := addrs[0]
-	oldLeader := c.GetLeaderAddr()
+	oldLeader := c.getLeaderAddr()
 	if addr == oldLeader {
 		return nil
 	}
 
-	if _, err := c.getOrCreateGRPCConn(addr); err != nil {
+	if _, err := c.GetOrCreateGRPCConn(addr); err != nil {
 		log.Warn("[pd] failed to connect leader", zap.String("leader", addr), errs.ZapError(err))
 		return err
 	}
 	// Set PD leader and Global TSO Allocator (which is also the PD leader)
 	c.leader.Store(addr)
-	c.allocators.Store(globalDCLocation, addr)
-	c.scheduleUpdateTokenConnection()
```

Comment:
Previously, we only run `scheduleUpdateTokenConnection`. But now, we do both `scheduleCheckTSODispatcher` and `scheduleUpdateTokenConnection`.

Reply:
As the same answer above. Also pasted below:
I didn't remove. Moved to switchLeader actually. We registered a leaderSwitchedCallback which calls c.scheduleCheckTSODispatcher(). Inside of c.switchLeader(), when the leader switched succeeds, all the leaderSwitchedCallbacks will be invoked, including c.scheduleCheckTSODispatcher() and c.scheduleUpdateTokenConnection(). The refactor doesn't change any logic.

Reply:
The refactor didn't intend to change any logic, but I indeed find there is difference introduced by my change. scheduleUpdateTokenConnection is called when switchLeader() returns nil which have two cases: 
1. the addr switched to is equal to the existing leader, so no leader switching happens in this case; L376
2. successfully switched leader to another host

With my change, if addr == oldLeader L376,   scheduleUpdateTokenConnection isn't called. ** if there is no change, do we need to tryResourceManagerConnect? My understanding is no. @JmPotato to confirm**
After leader successfully being switched, both scheduleUpdateTokenConnection and scheduleCheckTSODispatcher are invoked.

Reply:
@JmPotato @rleungx Could you please help to double check I**f there is no leader change, do we need to tryResourceManagerConnect?**

Reply:
There is no need to reconnect the resource manager if the leader remains the same.

Reply:
Thanks for the confirmation! Then my refactor should be ok, though it changed the previous logic that tryResourceManagerConnect was triggered even the leader kept the same.

Reply:
As mentioned above in another thread (also pasted below), I have moved scheduleCheckTSODispatcher back.

_I shouldn't move. After it's moved to inside of switchLeader(), it won't be called if any local tso allocator leader changed, which caused some tests to fail. I have added the equivalent logic back in this commit https://github.com/tikv/pd/commit/6ce5806648f3c416f33f7c01ffae0205f8f902fb._